py/server/deephaven/table.py

```diff
@@ -62,6 +65,86 @@ def _j_py_script_session() -> _JPythonScriptSession:
         return None
 
 
+def _encode_signature(fn: Callable) -> str:
+    """Encode the signature of a Python function by mapping the annotations of the parameter types and the return
+    type to numpy dtype chars (i,l,h,f,d,b,?,O), and pack them into a string with parameter type chars first,
+    in their original order, followed by the delimiter string '->', then the return type_char.
+
+    If a parameter or the return of the function is not annotated, the default 'O' - object type, will be used.
+    """
+    try:
+        sig = inspect.signature(fn)
+    except:
+        return ""
+
+    parameter_types = []
+    for n, p in sig.parameters.items():
+        try:
+            np_dtype = np.dtype(p.annotation if p.annotation else "object")
+            parameter_types.append(np_dtype)
+        except TypeError:
+            parameter_types.append(np.dtype("object"))
+
+    try:
+        return_type = np.dtype(sig.return_annotation if sig.return_annotation else "object")
+    except TypeError:
+        return_type = np.dtype("object")
+
+    np_type_codes = [np.dtype(p).char for p in parameter_types]
+    np_type_codes.extend(["-", ">", np.dtype(return_type).char])
+    return "".join(np_type_codes)
+
+
+def dh_vectorize(fn: Callable) -> Callable:
+    """A decorator to vectorize a Python function used in Deephaven query formulas and invoked on a row basis.
+
+    The result function will accept arrays of the declared parameter data types (vs. scalar) and iterate through them
+    together and call the wrapped function on each iteration and pack the result of each call into an array and return
+    the result array upon exhausting array parameters.
+    """
+    sig_code = _encode_signature(fn)
+    return_type = sig_code[-1]
+
+    @wraps(fn)
+    def wrapper(*args):
+        if len(args) != len(sig_code) - len("->?"):
+            raise ValueError(f"The number of arguments doesn't match the function signature. {len(args)}, {sig_code}")
+
+        def is_chunked(v) -> bool:
+            if isinstance(v, str) or not isinstance(v, jpy.JType):
+                return False
+            try:
+                iter(v)
+                return True
+            except:
+                return False
```

Comment:
Can you add a comment about how this works and why it is the right way to do it?

Is there any way this could trigger on list, array, or other inputs, when it shouldn't?